File: java/client/src/net/kio/its/client/SocketWorker.java
package net.kio.its.client;

import net.kio.its.client.events.*;
import net.kio.its.event.Event;
import net.kio.its.logger.LogType;
import net.kio.its.responsesystem.IResponseWorker;
import net.kio.its.responsesystem.Response;
import net.kio.its.responsesystem.ResponseManager;
import net.kio.security.dataencryption.EncryptedRequestManager;
import net.kio.security.dataencryption.KeysGenerator;

import java.io.*;
import java.net.Socket;
import java.net.SocketException;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.UUID;

public class SocketWorker extends Thread implements IResponseWorker {

    private final Client client;

    private final String targetIp;
    private final int targetPort;
    private Socket socket;
    private boolean isConnected;
    private boolean isAuthenticated;

    private DataOutputStream dataOutputStream;
    private DataInputStream dataInputStream;

    private final KeysGenerator keysGenerator;
    private final ResponseManager responseManager;

    private String commandPrefix;
    private String requestSeparator;

    private boolean connectionProtocol;

    public SocketWorker(Client client, String targetIp, int targetPort) {
        this.client = client;
        this.targetIp = targetIp;
        this.targetPort = targetPort;
        this.keysGenerator = new KeysGenerator();
        this.responseManager = new ResponseManager();
        setName("Socket Worker #" + getId());
    }

    public void startWorker() {
        try {
            connectSocket();
            start();
        } catch (IOException exception) {
            exception.printStackTrace();
        }
    }

    public void stopWorker() {
        if (isConnected) {
            sendCommand("disconnect");
            interrupt();
        }
    }

    public void connectSocket() throws IOException {
        if (socket != null && isConnected) throw new SocketException("Socket is already connected");
        else {
            socket = new Socket(targetIp, targetPort);
            socket.setSoTimeout(0);
            dataOutputStream = new DataOutputStream(socket.getOutputStream());
            dataInputStream = new DataInputStream(socket.getInputStream());
            isConnected = true;
        }
    }

    public void disconnectSocket() throws IOException {
        if (socket == null) throw new SocketException("Socket isn't connected");
        else {
            stopWorker();
            socket.close();
            isConnected = false;
            interrupt();
            socket = null;
        }
    }

    @Override
    public void run() {
        handleServerSocket();
    }

    private void handleServerSocket() {
        try {
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(dataInputStream));
            startConnectionProtocol();
            String line;
            while (!Thread.interrupted() && (line = bufferedReader.readLine()) != null) {
                onLineRead(line);
            }
            disconnectSocket();
        } catch (SocketException e) {
            client.getLogger().log(LogType.CRITICAL, getName() + " - Unlegit disconnection from server " + targetIp + " on port " + targetPort);
            ServerSocketClosedEvent event = new ServerSocketClosedEvent(socket, this);
            client.getEventsManager().callEvent(event);
            interrupt();
        } catch (IOException exception) {
            exception.printStackTrace();
        }
    }

    private void onLineRead(String line) {

        String decryptedData;
        if (keysGenerator.getSecretKey() != null && (decryptedData = EncryptedRequestManager.decrypt(line, keysGenerator)) != null) {
            onDataReceived(new String[]{decryptedData, line}, true);
        } else {
            onDataReceived(new String[]{line}, false);
        }


    }

    private void onDataReceived(String[] data, boolean encrypted) {

        UUID msgUUID;

        if (data[0].startsWith("response:")) {
            msgUUID = UUID.fromString(data[0].substring(9, 45));
            data[0] = data[0].substring(45);
            client.getLogger().log(LogType.DEBUG, getName() + " - Response received for UUID " + msgUUID);
            responseManager.onResponseReceived(msgUUID, data[0]);
        } else {
            msgUUID = UUID.fromString(data[0].substring(0, 36));
            data[0] = data[0].substring(36);
        }

        if (commandPrefix != null && data[0].startsWith(commandPrefix)) {
            onCommandReceived(data[0], msgUUID, encrypted);
        } else {
            client.getLogger().log(LogType.DEBUG, getName() + " - Data received (encrypted=" + encrypted + ", uuid=" + msgUUID + ") : " + data[0]);

            Event event = new RawDataReceivedEvent(this, data[0], msgUUID);
            if (encrypted) {
                event = new EncryptedDataReceivedEvent(this, data[1], data[0], msgUUID);
            }
            client.getEventsManager().callEvent(event);


            if (connectionProtocol) {
                if (!onConnectionProtocolDataReceived(data[0])) {
                    client.getLogger().log(LogType.CRITICAL, "Unable to connect to the server");
                    stopWorker();
                }
            }


        }

    }

    private void onCommandReceived(String command, UUID msgUUID, boolean encrypted) {
        client.getLogger().log(LogType.DEBUG, getName() + " -  Command Received (encrypted=" + encrypted + ", uuid=" + msgUUID + ") : " + command);
        CommandReceivedEvent event = new CommandReceivedEvent(this, command, commandPrefix, requestSeparator, msgUUID);
        client.getEventsManager().callEvent(event);
    }


    private void startConnectionProtocol() {
        connectionProtocol = true;
        sendData("version:" + Client.getVersion(), false);
    }

    private boolean onConnectionProtocolDataReceived(String data) {

        if (data.startsWith("version:")) {
            boolean version = Boolean.parseBoolean(data.substring(8));
            if (!version) return false;
            else {
                sendData("macAddress:" + Client.getMacAddress(), false);
            }
        } else if (data.startsWith("publicKey:")) {
            String array = data.replace("publicKey:", "");
            try {
                keysGenerator.setPublicKey(KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(Base64.getDecoder().decode(array))));
                keysGenerator.generateKeys(true, false);
                sendData("secretKey:" + EncryptedRequestManager.encryptSecretKey(keysGenerator), false);
            } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
                e.printStackTrace();
                return false;
            }

        } else if (data.startsWith("separator:")) {
            requestSeparator = data.replace("separator:", "");
        } else if (data.startsWith("commandPrefix:")) {
            commandPrefix = data.replace("commandPrefix:", "");
        } else if (data.startsWith("connected:")) {
            boolean connected = Boolean.parseBoolean(data.replace("connected:", ""));
            isAuthenticated = connected;
            connectionProtocol = !connected;
            if (connected) {
                ConnectionProtocolSuccessEvent event = new ConnectionProtocolSuccessEvent(this);
                client.getEventsManager().callEvent(event);
            }
            return connected;
        } else {
            return false;
        }

        return true;
    }

    public Response sendData(String data) {
        return sendData(data, (UUID) null, true);
    }

    public Response sendData(String data, boolean encrypt) {
        return sendData(data, (UUID) null, encrypt);
    }

    public Response sendData(String data, String responseUUID) {
        return sendData(data, UUID.fromString(responseUUID), true);
    }

    public Response sendData(String data, String responseUUID, boolean encrypt) {
        return sendData(data, UUID.fromString(responseUUID), encrypt);
    }

    public Response sendData(String data, UUID responseUUID, boolean encrypt) {
        String rawData = data;
        Response response = new Response(this, responseUUID);
        OutputStreamWriter outputStreamWriter = new OutputStreamWriter(dataOutputStream);
        PrintWriter printWriter = new PrintWriter(outputStreamWriter, true);
        data = (responseUUID != null ? "response:" : "") + response.getMsgUUID() + data;
        if (encrypt) {
            data = EncryptedRequestManager.encrypt(data, keysGenerator);
        }
        printWriter.println(data);
        client.getLogger().log(LogType.DEBUG, getName() + " - Data sent (encrypted=" + encrypt + ", uuid=" + response.getMsgUUID() + ") : " + rawData);
        return response;
    }

    public Response sendCommand(String command, String... args) {
        return sendData(commandPrefix + command + requestSeparator + String.join(requestSeparator, args));
    }

    public boolean isConnected() {
        return isConnected;
    }

    public boolean isAuthenticated() {
        return isAuthenticated;
    }

    public String getTargetIp() {
        return targetIp;
    }

    public int getTargetPort() {
        return targetPort;
    }

    public ResponseManager getResponseManager() {
        return responseManager;
    }
}

File: client/socket_worker.go
package client

import (
	"errors"
	"fmt"
	"net"
	"strconv"
	"strings"
	"time"

	"github.com/TorKhanNetwork/Networking/data_encryption"
	"github.com/TorKhanNetwork/Networking/events_system"
	"github.com/TorKhanNetwork/Networking/response_system"
	"github.com/google/uuid"
	"github.com/kataras/golog"
)

const INPUT_BUFFER_SIZE = 65536

type SocketWorker struct {
	Id                                           int
	client                                       *Client
	targetIp, commandPrefix, requestSeparator    string
	targetPort                                   int
	connection                                   *net.TCPConn
	connected, authenticated, connectionProtocol bool
	keysGenerator                                data_encryption.KeysGenerator
	responseManager                              response_system.ResponseManager
	quitChan                                     chan bool
}

func NewSocketWorker(id int, client *Client, targetIp string, targetPort int) SocketWorker {
	return SocketWorker{
		Id:              id,
		client:          client,
		targetIp:        targetIp,
		targetPort:      targetPort,
		keysGenerator:   data_encryption.NewGenerator(),
		responseManager: response_system.NewResponseManager(),
	}
}

func (socketWorker *SocketWorker) GetName() string {
	return fmt.Sprintf("  |  %s  |  SocketWorker #%d ", socketWorker.client.Name, socketWorker.Id)
}

func (socketWorker *SocketWorker) GetResponseManager() response_system.ResponseManager {
	return socketWorker.responseManager
}

func (socketWorker *SocketWorker) StartWorker() (err error) {
	if err = socketWorker.ConnectSocket(); err != nil {
		return err
	}
	socketWorker.quitChan = make(chan bool)
	go socketWorker.handleServerSocket()
	return
}

func (socketWorker *SocketWorker) StopWorker() {
	if socketWorker.connected {
		socketWorker.client.RemoveSocketWorker(*socketWorker)
		socketWorker.quitChan <- true
	}
}

func (socketWorker *SocketWorker) ConnectSocket() error {
	if socketWorker.connected {
		return errors.New("socket is already connected")
	}
	tcpAddr, err := net.ResolveTCPAddr("tcp", socketWorker.targetIp+":"+strconv.Itoa(socketWorker.targetPort))
	if err != nil {
		return fmt.Errorf("unable to resolve TCP Address : %s", err)
	}
	connection, err := net.DialTCP("tcp", nil, tcpAddr)
	if err != nil {
		return fmt.Errorf("unable to connect the socket : %s", err)
	}
	if connection == nil {
		return errors.New("connection to the socket is nil")
	}
	connection.SetDeadline(time.Time{})
	socketWorker.connection = connection
	socketWorker.connected = true
	return nil
}

func (socketWorker *SocketWorker) DisconnectSocket() error {
	if !socketWorker.connected {
		return errors.New("socket isn't connected")
	}
	socketWorker.SendData("disconnect", uuid.Nil, socketWorker.authenticated)
	err := socketWorker.connection.Close()
	if err != nil {
		return fmt.Errorf("unable to close socket : %s", err)
	}
	socketWorker.connected = false
	socketWorker.StopWorker()
	return nil
}

func (socketWorker *SocketWorker) handleServerSocket() {
	buf := make([]byte, INPUT_BUFFER_SIZE)
	socketWorker.startConnectionProtocol()
	for {
		mLen, err := socketWorker.connection.Read(buf)
		if err != nil {
			break
		}
		select {
		case <-socketWorker.quitChan:
			if err := socketWorker.DisconnectSocket(); err != nil {
				golog.Errorf("%s - %s", socketWorker.GetName(), err)
			}
			return
		default:
			for _, line := range strings.Split(string(buf[:mLen]), "\n") {
				socketWorker.onLineRead(line)
			}
		}
	}
	var e events_system.Event = NewServerSocketDisconnectEvent(socketWorker)
	socketWorker.client.EventsManager.CallEvent(&e)
}

func (socketWorker *SocketWorker) onLineRead(line string) {
	line = strings.TrimSpace(line)
	if line != "" {
		if decryptedData, err := data_encryption.Decrypt(line, socketWorker.keysGenerator); err == nil {
			socketWorker.onDataReceived([]string{decryptedData, line}, true)
		} else {
			socketWorker.onDataReceived([]string{line}, false)
		}
	}

}

func (socketWorker *SocketWorker) onDataReceived(data []string, encrypted bool) {
	var msgUUID uuid.UUID
	var err error
	if strings.HasPrefix(data[0], "response:") {
		msgUUID, err = uuid.Parse(data[0][9:45])
		if err != nil {
			golog.Errorf("%s - Unable to parse response UUID : %s", socketWorker.GetName(), err)
			return
		}
		data[0] = data[0][45:]
		golog.Debugf("%s - Response received for UUID %s", socketWorker.GetName(), msgUUID)
		go socketWorker.responseManager.OnResponseReceived(msgUUID, data[0])
	} else {
		var err error
		msgUUID, err = uuid.Parse(data[0][:36])
		if err != nil {
			golog.Errorf("%s - Unable to parse response UUID : %s", socketWorker.GetName(), err)
			return
		}
		data[0] = data[0][36:]
	}

	if socketWorker.connectionProtocol {
		golog.Debugf("%s - Data received (encrypted=%t, uuid=%s) : %s", socketWorker.GetName(), encrypted, msgUUID, data[0])
		if !socketWorker.onConnectionProtocolDataReceived(data[0]) {
			golog.Default.Errorf("%s - Unable to authenticate to the server", socketWorker.GetName())
		}
	} else {
		if socketWorker.commandPrefix != "" && strings.HasPrefix(data[0], socketWorker.commandPrefix) {
			socketWorker.onCommandReceived(data[0], msgUUID, encrypted)
		} else {
			golog.Debugf("%s - Data received (encrypted=%t, uuid=%s) : %s", socketWorker.GetName(), encrypted, msgUUID, data[0])
			if data[0] == "disconnect" {
				socketWorker.StopWorker()
				var e events_system.Event = NewServerDisconnectEvent(socketWorker)
				socketWorker.client.EventsManager.CallEvent(&e)
			}
		}
	}

}

func (socketWorker *SocketWorker) onCommandReceived(command string, msgUUID uuid.UUID, encrypted bool) {
	golog.Debugf("%s - Command Received (encrypted=%t, uuid=%s) : %s", socketWorker.GetName(), encrypted, msgUUID, command)
	var event events_system.Event = NewCommandReceivedEvent(socketWorker, command, socketWorker.commandPrefix, socketWorker.requestSeparator, msgUUID)
	socketWorker.client.EventsManager.CallEvent(&event)
}

func (socketWorker *SocketWorker) startConnectionProtocol() {
	socketWorker.connectionProtocol = true
	socketWorker.SendData("version:"+CLIENT_VERSION, uuid.Nil, false)
}

func (socketWorker *SocketWorker) onConnectionProtocolDataReceived(data string) bool {
	if strings.HasPrefix(data, "version:") {
		if version := strings.ToLower(data[8:]) == "true"; version {
			socketWorker.SendData("macAddress:"+socketWorker.client.MacAddress, uuid.Nil, false)
		} else {
			return false
		}
	} else if strings.HasPrefix(data, "publicKey:") {
		pub, err := ParseRsaPublicKeyFromPemStr(data[10:])
		if err != nil {
			golog.Errorf("%s - Failed to parse server public key : %s", socketWorker.GetName(), err)
			return false
		}
		socketWorker.keysGenerator.PublicKey = *pub
		socketWorker.keysGenerator.GenerateKeys(true, false)
		secretKey, err := data_encryption.EncryptSecretKey(socketWorker.keysGenerator)
		if err != nil {
			golog.Errorf("%s - Unable to encrypt secret key : %s", socketWorker.GetName(), err)
			return false
		}
		socketWorker.SendData("secretKey:"+secretKey, uuid.Nil, false)
	} else if strings.HasPrefix(data, "commandPrefix:") {
		socketWorker.commandPrefix = data[14:]
	} else if strings.HasPrefix(data, "connected:") {
		connected := strings.ToLower(data[10:]) == "true"
		socketWorker.authenticated = connected
		socketWorker.connectionProtocol = !connected
		if connected {
			var e events_system.Event = NewConnectionProtocolSuccessEvent(socketWorker)
			socketWorker.client.EventsManager.CallEvent(&e)
		}
		return connected
	} else {
		return false
	}
	return true
}

func (socketWorker *SocketWorker) SendData(data string, responseUUID uuid.UUID, encrypt bool) *response_system.Response {
	rawData := data
	response := response_system.NewResponse(socketWorker, responseUUID)
	data = response.UUID.String() + data
	if responseUUID != uuid.Nil {
		data = "response:" + data
	}
	if encrypt {
		var err error
		data, err = data_encryption.Encrypt(data, socketWorker.keysGenerator)
		if err != nil {
			golog.Errorf("%s - Unable to encrypt data : %s", socketWorker.GetName(), err)
			return nil
		}
	}
	_, err := socketWorker.connection.Write([]byte(data + "\n"))
	if err != nil {
		golog.Errorf("%s - Unable to send data : %s", socketWorker.GetName(), err)
		return nil
	}
	golog.Debugf("%s - Data sent (encrypted=%t, uuid=%s) : %s", socketWorker.GetName(), encrypt, response.UUID, rawData)
	return &response
}

func (socketWorker *SocketWorker) SendCommand(command string, args ...string) *response_system.Response {
	data := socketWorker.commandPrefix + command
	for _, arg := range args {
		arg = strings.ReplaceAll(arg, "\"", "\\\"")
		data += " \"" + arg + "\""
	}
	return socketWorker.SendData(data, uuid.Nil, true)
}
